mantidproject/mantid
# Improve Precompiled header files to speed up Windows build

**Description of work.**

Improved the coverage of the precompiled headers configuration to speed up Windows build times. This will help developers and will also speed up the Jenkins Windows build time. Specifically:

- Add some extra headers to various PrecompiledHeader.h files across Framework, Qt5, Unit tests
- Turn on precompiled headers for the PythonInterface targets
- Fixed problem with Precompiled Headers cmake set up for unit tests which meant they weren't active at all
- Made a couple of code changes to add qualifiers to variables to ensure code still compiles. Activation of precompiled headers changes order of includes in some cases so some variables became ambiguous. 

I've focussed on 3rd party headers plus a small number of Mantid headers that have a large impact. Inclusion of headers in here has a "cost" that needs to be weighed up against the "win" from the speed up:
- the precompiled header gets automatically added to all .cpp units so it can result in a compilation failure on a non-Windows platform if a Windows developer inadvertently relies on a precompiled header
- if a Mantid header is included in a target's PrecompiledHeader.h and that Mantid header changes, it triggers a rebuild of the whole target. This could reduce the number of incremental builds that happen on your local system and in the Windows Jenkins jobs eg changing MantidAPI/Algorithm.h will cause rebuild of Algorithm target. Ironically, the the changes in this PR to various PrecompiledHeader.h files will also trigger more full rebuilds on the first Jenkins job on each node once this hits master. But that will be a one-off effect. You can see this in the Windows Jenkins job for this PR which was longer than usual at (~38 mins vs ~30 mins)

For some reason I didn't see any speed up at all from adding the C++ standard library headers eg \<vector>, \<map> etc. This is strange because most online examples of precompiled headers include things like \<vector>. Maybe Visual Studio already precompiles these??

I didn't find the Build\Rebuild options in Visual Studio very helpful for measuring the compilation time of a specific target eg Algorithms. The Rebuild option rebuilt all of the dependencies rather than just the .cpp files in Algorithms. So to get some timings I deleted all the .obj files from build\Framework\Algorithms\Algorithms.dir\Debug and then ran Build. While this would normally do an incremental build, the absence of all the .obj files means it does a rebuild of the Algorithms .cpp files

Here are the speed ups I see on my laptop for Debug builds:
- Framework: down from 8m30s to 5m20s
- FrameworkTests (where precompiled headers were completely off): 13m50s to 7m10s

**To test:**

Ensure the USE_PRECOMPILED_HEADERS cmake setting is set to "on"
Rebuild the Framework group of projects in Visual Studio before and after these changes and observe a speed up. I suggest you do a Debug build because most development work will use that. On my laptop the build times reduces from about 8m30s to 5m20s

Fixes #30401. 

*This does not require release notes* because **it has no functional impact**

---

#### Reviewer ####

Please comment on the following ([full description](http://developer.mantidproject.org/ReviewingAPullRequest.html)):

##### Code Review #####

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there is GUI work does it follow the [GUI standards](http://developer.mantidproject.org/Standards/GUIStandards.html)?
- If there are changes in the release notes then do they describe the changes appropriately?

##### Functional Tests #####

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.
